### ISSUE TITLE
Ceph Teardown - Deleting the data on hosts

### ISSUE THREAD

**Issue by aperullo** [rook/rook#3646]
The [Ceph teardown ](https://github.com/rook/rook/blob/master/Documentation/ceph-teardown.md#delete-the-data-on-hosts) docs say that `In the future this step will not be necessary when we build on the K8s local storage feature.` 

What is the status of this and what does this mean exactly, and how would one utilize this to not have to manually scrub the filesystem?

**Comment by travisn** [comment]
@aperullo That comment was based on the feature described in #919 where the OSDs would be backed by PVs. This feature just merged a few days ago, so the comment you found in the docs can be clarified now. 

If the OSDs are backed by PVs, the cleanup of the PVs would be dependent on the PV reclaim policy. In theory you could just delete the PVs and they would be cleaned up by automatically by their storage provisioner. For example, if running in aws, the gp2 provisioner would remove the ebs PVs and there would be no further cleanup. Other provisioners may not behave quite as smoothly though. For example, the local storage provisioner might clean the device, though I'm not sure if an additional cleanup step would be needed.

**Comment by aperullo** [comment]
Oh well that is good to hear. Still, reading through that doesn't give me a clear idea of exactly what that pr means for the manual clean up step. How would I configure my rook storage class to have this automatic cleanup? Thanks

**Comment by travisn** [comment]
There are two different storage classes to consider:
1. The Rook storage class, which allows you to consume the Ceph storage. This will be the storage classes in rook documentation, such as [here](https://rook.io/docs/rook/master/ceph-block.html). During cleanup, the pods consuming the rook storage class should be removed before Rook is deleted. Otherwise, your pods will hang trying to connect to the Ceph OSDs which would be removed.
1. The storage class from which OSD storage is provisioned, which is a new option with #919. If running in AWS, this might be the gp2 provisioner. If on premise it might be the local-storage provider, and so on. When you delete the Rook cluster CR, the PVCs backing the OSDs would be deleted. The PVs should then be deleted by the owner of that provisioner. Again, if in AWS this means the gp2 provider would delete the ebs-based PVs. In AWS the PV cleanup happens automatically from what i've seen. 

Does that help clarify?

**Comment by aperullo** [comment]
Ah I see, point 1 is clear. I didn't realize there were two things called storage class.

My confusion with point 2 probably stems from my lack of familiarity with how rook works under the hood. The field that the PR adds support for is for the custom resource, so is the configuration for it [this field?](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/cluster-on-pvc.yaml#L30). I think I found the documentation to go with the PR. And I think some further reading is really just in order. Thanks again for your help.

**Comment by travisn** [comment]
> The field that the PR adds support for is for the custom resource, so is the configuration for it this field?

Correct, point 2 is based on the volume claim templates that you referenced.

**Comment by aperullo** [comment]
I can't seem to get that example working based on the new documentation [here](https://github.com/rook/rook/blob/master/Documentation/ceph-cluster-crd.md#storage-class-device-sets)



```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook
spec:
  cephVersion:
    image: ceph/ceph:v14.2.1-20190430
    allowUnsupported: false
  dataDirHostPath: /var/lib/rook
  mon:
    count: 3
    allowMultiplePerNode: false
    volumeClaimTemplate:
      spec:
        storageClassName: local-storage
        resources:
          requests:
            storage: 10Gi
  dashboard:
    enabled: true
  network:
    # toggle to use hostNetwork
    hostNetwork: false
  rbdMirroring:
    workers: 0
  annotations:
  #resources:
  # The requests and limits set here, allow the mgr pod to use half of one CPU core and 1 gigabyte of memory
  #    mgr:
  #      limits:
  #        cpu: "500m"
  #        memory: "1024Mi"
  #      requests:
  #        cpu: "500m"
  #        memory: "1024Mi"
  # The above example requests/limits can also be added to the mon and osd components
  #    mon:
  #    osd:
  storage: # cluster level storage configuration and selection
    useAllNodes: true
    useAllDevices: true
    deviceFilter:
    location:
    config:
    #- path: /var/lib/rook
    storageClassDeviceSets:
    - name: set1
      count: 6
      volumeClaimTemplates:
      - metadata:
          name: data
        spec:
          resources:
            requests:
              storage: 30Gi
          storageClassName: local-storage
          volumeMode: Block
          accessModes:
            - ReadWriteOnce
```

The mgr is reporting 

```
debug 2019-08-21 19:13:32.298 7ff256eee700  1 mgr.server send_report Not sending PG status to monitor yet, waiting for OSDs
debug 2019-08-21 19:13:34.298 7ff256eee700  1 mgr.server send_report Not sending PG status to monitor yet, waiting for OSDs
debug 2019-08-21 19:13:36.298 7ff256eee700  1 mgr.server send_report Giving up on OSDs that haven't reported yet, sending potentially incomplete PG state to mon
debug 2019-08-21 19:13:36.298 7ff256eee700  0 log_channel(cluster) log [DBG] : pgmap v10: 300 pgs: 300 unknown; 0 B data, 0 B used, 0 B / 0 B avail
debug 2019-08-21 19:13:38.299 7ff256eee700  0 log_channel(cluster) log [DBG] : pgmap v13: 400 pgs: 400 unknown; 0 B data, 0 B used, 0 B / 0 B avail
debug 2019-08-21 19:13:40.299 7ff256eee700  0 log_channel(cluster) log [DBG] : pgmap v16: 400 pgs: 400 unknown; 0 B data, 0 B used, 0 B / 0 B avail
debug 2019-08-21 19:13:42.300 7ff256eee700  0 log_channel(cluster) log [DBG] : pgmap v18: 400 pgs: 400 unknown; 0 B data, 0 B used, 0 B / 0 B avail
```
Only after this did the OSD-prepare jobs show up. This makes it seem like the mgr isn't waiting long enough for the osd's to appear but I'm really not sure. It seems to work fine if I'm not using the new feature. Any idea whats going on?

**Comment by travisn** [comment]
> Only after this did the OSD-prepare jobs show up. This makes it seem like the mgr isn't waiting long enough for the osd's to appear but I'm really not sure. It seems to work fine if I'm not using the new feature. Any idea whats going on?

The OSDs are created after the mgr is started, so you can ignore those messages in the mgr until the OSDs come up. Or maybe I'm missing what error you're referring to.

> could it be that the helm chart isn't updated yet? As thats what I'm using to set up the operator.

The helm chart only configures the operator. Everything related to the Ceph cluster is really in the cluster.yaml, which is not part of the helm chart.

> do I need to deploy the local storage external provisioner if I dont want to have to clean up the volumes myself?

Yes, I believe that's true, but I can't say I've tested the local provisioner myself.
